src/crimson/tools/kvstore/crimson_objectstore_tool.cc

```diff
@@ -0,0 +1,517 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab ft=cpp
+
+#include <boost/program_options/value_semantic.hpp>
+#include <boost/program_options/variables_map.hpp>
+#include <boost/program_options/parsers.hpp>
+#include <cstdlib>
+#include <cstring>
+#include <fmt/format.h>
+#include <fmt/ostream.h>
+#include <fstream>
+#include <iostream>
+#include <memory>
+#include <optional>
+#include <ostream>
+#include <seastar/core/app-template.hh>
+#include <seastar/core/signal.hh>
+#include <string>
+#include <string_view>
+#include <vector>
+#include <sys/stat.h>
+
+#include "common/Formatter.h"
+#include "common/hobject.h"
+#include "crimson/common/errorator.h"
+#include "crimson/osd/stop_signal.h"
+#include "crimson/osd/osd_meta.h"
+
+#include "objectstore_tool.h"
+#include "osd/osd_types.h"
+#include "seastar/core/thread.hh"
+#include "crimson/common/config_proxy.h"
+#include "crimson/os/futurized_collection.h"
+#include "crimson/os/futurized_store.h"
+#include "seastar/util/closeable.hh"
+#include "seastar/util/log.hh"
+#include "crimson/os/seastore/segment_manager.h"
+
+namespace bpo = boost::program_options;
+using crimson::common::sharded_conf;
+using crimson::common::local_conf;
+
+struct objectstore_config_t {
+  std::string data_path;
+  std::string store_type;
+  std::string pg;
+  unsigned int shard_id = 0;
+  std::string object;
+  std::string omap_start;
+  std::string omap_key;
+  std::string omap_value;
+  std::string object_namespace;
+  std::string format = "json-pretty";
+  bool debug = false;
+  bool list_pgs = false;
+  bool list_objects = false;
+  bool list_omap = false;
+  bool get_omap = false;
+  bool set_omap = false;
+  bool remove_omap = false;
+
+  void populate_options(bpo::options_description &desc) {
+    desc.add_options()
+      ("data-path",
+       bpo::value<std::string>(&data_path)->required(),
+       "osd data path (required)")
+      ("store-type",
+       bpo::value<std::string>(&store_type)->required(),
+       "store type, e.g seastore (required)")
+      ("debug",
+       bpo::bool_switch(&debug),
+       "set logger level as debug")
+      ("list-pgs",
+       bpo::bool_switch(&list_pgs),
+       "list all pgs")
+      ("list-objects",
+       bpo::bool_switch(&list_objects),
+       "list all objects in a pg")
+      ("list-omap",
+       bpo::bool_switch(&list_omap),
+       "list all omap keys in an object")
+      ("get-omap",
+       bpo::bool_switch(&get_omap),
+       "get specific omap value by key")
+      ("set-omap",
+       bpo::bool_switch(&set_omap),
+       "set omap key-value for an object")
+      ("remove-omap",
+       bpo::bool_switch(&remove_omap),
+       "remove omap key from an object")
+      ("pg",
+       bpo::value<std::string>(&pg)->default_value(""),
+       "pg name")
+      ("shard-id",
```

Comment:
See `FuturizedStore::list_collections()`, seems to me `shard-id` is statically associated with the `pg` in query, which is decided by crimson OSD, we cannot assign it manually in the tool.